plotly/widgets/graphWidget.js

```diff
@@ -0,0 +1,105 @@
+
+            // variable plotly_domain in the case of enterprise
+            var url_parts = _graph_url.split('/');
+            var plotly_domain = url_parts[0] + '//' + url_parts[2];
+
+            // Place IFrame in output cell div `$el`
+            that.$el.css('width', '100%');
+            that.$graph = $(['<iframe id="'+frameId+'"',
+                             'src="'+_graph_url+'.embed"',
+                             'seamless',
+                             'style="border: none;"',
+                             'width="100%"',
+                             'height="600">',
+                             '</iframe>'].join(' '));
+            that.$graph.appendTo(that.$el);
+
+            that.$loading = $('<div id="'+loadingId+'">Initializing...</div>')
+                            .appendTo(that.$el);
+
+            // initialize communication with the iframe
+            if(!('pingers' in window)){
+                window.pingers = {};
+            }
+
+            window.pingers[frameId] = setInterval(function() {
+                that.graphContentWindow = $('#'+frameId)[0].contentWindow;
+                that.graphContentWindow.postMessage({ping: true}, plotly_domain);
+            }, 200);
+
+            // Assign a message listener to the 'message' events
+            // from iframe's postMessage protocol.
+            // Filter the messages by iframe src so that the right message
+            // gets passed to the right widget
+            if(!('messageListeners' in window)){
+                 window.messageListeners = {};
+            }
+
+            window.messageListeners[frameId] = function(e) {
+                if(_graph_url.indexOf(e.origin)>-1) {
+                    var frame = document.getElementById(frameId);
+
+                    if(frame === null){
+                        // frame doesn't exist in the dom anymore, clean up it's old event listener
+                        window.removeEventListener('message', window.messageListeners[frameId]);
+                        clearInterval(window.pingers[frameId]);
+                    } else if(frame.contentWindow === e.source) {
+                        // TODO: Stop event propagation, so each frame doesn't listen and filter
+                        var frameContentWindow = $('#'+frameId)[0].contentWindow;
+                        var message = e.data;
+
+                        if(message==='pong') {
+                            $('#loading-'+frameId).hide();
+                            clearInterval(window.pingers[frameId]);
+                            that.send({event: 'pong', graphId: frameId});
+                        } else if (message.type==='hover' ||
+                                   message.type==='zoom'  ||
+                                   message.type==='click' ||
+                                   message.type==='unhover') {
+                            that.send({event: message.type, message: message, graphId: frameId});
```

Comment:
One thing that we might want to add is the option of a user-defined `preprocessor` that filters or transforms the message before it's sent.

For `click` events, the entire graph JSON is sent in this message. For large JSONs, this actually takes a while to pass into the IPython kernel, making widgets like this [IPython contour plotter](https://camo.githubusercontent.com/efe97aa899832ab7e8466775e6085d33bc3553cb/687474703a2f2f692e696d6775722e636f6d2f6d754b577649362e676966) pretty slow.